### PR TITLE
add a wrapper around Mir with an exit node, dominators returns error when nodes are unreachable

### PR DESCRIPTION
**Pull request by scottcarr**
We want the Mir CFG to have an exit node to calculate post dominators.  The MirWithExit type allows us to add the exit node on demand.

When nodes are unreachable from the start node, dominators are undefined, so dominators now returns an error in that case.

### CODE REVIEW

**Comment by arielb1** [comment]
What is the motivation for postdoms here? LLVM seems to use them only for its region analysis.

**Comment by scottcarr** [comment]
> What is the motivation for postdoms here? LLVM seems to use them only for its region analysis.

For the "move up propagation" optimization I'm working on (which eliminates a temporary), it only fires if this particular use post dominates the temporary's definition. 

> BTW, this does not handle loops without an exit.

The way I was thinking to handle it is:

If the Mir CFG has loops without an exit, then there is no "the exit node" and the post-dominators are undefined.  When we calculate the dominators of the transposed the MirWithExit CFG, that graph will have unreachable nodes and dominators returns an error in that case.

**Comment by arielb1** [comment]
Use post-dominates definition? That means it won't work if there are panics in the middle, right?

**Comment by scottcarr** [comment]
> Use post-dominates definition?

I'm not sure what you mean?

> That means it won't work if there are panics in the middle, right?

Dominators::dominators doesn't panic when it encounters unreachable nodes, it returns an Result.  Callers should check the result if the graph might have unreachable nodes.  Mir's CFG shouldn't have unreachable nodes, AFAIK, but I can change Mir::dominators to return the Result if needed.

**Comment by arielb1** [comment]
@scottcarr 

The problem is that if you consider panic edges, your analysis will be reduced to be basically local, as every call has a panic edge which means that nothing post-dominates anything.

**Comment by scottcarr** [comment]
@arielb1 

Let me make sure I understand what you mean.  If we have:

```
bb0: {
  tmp0 = 5;
  tmp1 = 42;
  tmp2 = foo(tmp1) -> [return: bb1, unwind: bb2] 
}

bb1: {
  tmp3 = tmp0;
  ...
}

bb2: {
  resume;
}
``` 

You are suggesting we should optimize to:

```
bb0: {
  tmp3 = 5; // tmp0 optimized out
  tmp1 = 42;
  tmp2 = foo(tmp1) -> [return: bb1, unwind: bb2] 
}

bb1: {
  // tmp3 = tmp0 optimized out
  ...
}

bb2: {
  resume;
}
```

.. because "tmp3 = tmp0" from the first example post-dominates "tmp0 = 5" on all paths that do not end in a `resume;`?

**Comment by scottcarr** [comment]
FWIW, move up optimization does fire a non-zero number of times when building the compiler.  But it may be that all the statement pairs it optimizes are pretty local to each other.

https://github.com/rust-lang/rust/pull/34585

**Comment by nikomatsakis** [comment]
So I chatted a bit with @scottcarr on IRC. I don't think that panic edges are actually particularly relevant. I think that what it comes down to is that if you are going to move the write `B` so that it occurs at the point `A`:

```rust
B0: {
   TMP = ... // Point A
   ...
}

Bn: {
    X = TMP // Point B
    ...
}
```

then basically anything reachable from A without passing through B must not be able to observe the fact that `X` has changed before it was supposed to have changed. So, if you trace paths from `A` and you encounter a RETURN or UNWIND terminator (whatever we call those now), then you could conclude that because the local variable `X` is being popped, you can safely perform the optimization. But if you encounter some node that may observe the value of `X` (which might include calls, depending on whether the address of `X` has been taken and what kind of conservative rules we are using) then you can't safely move it backward.

In other words, I think @arielb1 is right that it might be better not to consider post-doms, but I think the focus on panics etc isn't that important.

**Comment by nikomatsakis** [comment]
(To be clear, I didn't read all the comments on this PR in depth.)

**Comment by arielb1** [comment]
I think we got this discussion totally wrong anyway. Here's my model of the optimization:

The optimization is to transform
```
S1:
    tmp = SRC
    ...
S2:
    DEST = tmp
```

to
```
S1:
    DEST = SRC
```

I think this is best split into 4 steps. I don't think we ever want to do the steps separately, but this clarifies the analysis rules.

Step 0 (original):
```
S1:
    tmp = SRC
    ...
S2:
    DEST = tmp
```

Step 1 - add additional write:
```
S1:
    tmp = SRC
    DEST = tmp
    ...
S2:
    DEST = tmp
```

This requires that `DEST` is dead at S1, and that it can be evaluated there (e.g. it is not a dereference of a pointer that is uninitialized there). Nothing else matters.

Step 2: separate out second temp
```
S1:
    tmp = SRC
    DEST = tmp
    ...
S2:
    tmp2 = DEST
    DEST = tmp2
```

This requires that the newly-added read links with the write of `tmp` - basically a "memdep" analysis.

This also requires that the address of `DEST` does not change, which is non-trivial because `DEST` can be the dereference of a pointer.

Step 3: remove S2

```
S1:
    tmp = SRC
    DEST = tmp
    ...
S2:
    NOP
```

After step 2, `S2` is obviously a NOP and can be removed.

Step 4: remove `tmp`

```
S1:
    DEST = SRC
S2:
    NOP
```

This is purely a local operation, but may require some sophistication if e.g. `SRC` is a function call. I think we are always justified doing it, but we should make sure it is OK.

**Comment by scottcarr** [comment]
Since we're not planning to use post dominators for move-up-propagation, should be close this PR and move discussion to https://github.com/rust-lang/rust/pull/34693?

**Comment by nikomatsakis** [comment]
@scottcarr I think we should.
